src/util/amount.rs

```diff
@@ -0,0 +1,754 @@
+            }
+            s = &s[1..];
+        }
+
+        let max_decimals = P::scale() - denom.scale();
+        let mut decimals = None;
+        let mut value: Inner = 0;
+        for c in s.as_bytes() {
+            match *c {
+                b'0'...b'9' => {
+                    // Do `value = 10 * value + digit`, catching overflows.
+                    match INNER_TEN.checked_mul(value) {
+                        None => return Err(ParseAmountError::TooBig),
+                        Some(n) => match n.checked_add((c - b'0').into()) {
+                            None => return Err(ParseAmountError::TooBig),
+                            Some(n) => value = n,
+                        },
+                    }
+                    // Increment the decimal digit counter if past decimal.
+                    decimals = match decimals {
+                        None => None,
+                        Some(d) if d == max_decimals => return Err(ParseAmountError::TooPrecise),
+                        Some(d) => Some(d + 1),
+                    }
+                }
+                b'.' => match decimals {
+                    None => decimals = Some(0),
+                    Some(_) => return Err(ParseAmountError::InvalidFormat),
+                },
+                _ => return Err(ParseAmountError::InvalidFormat),
+            }
+        }
+
+        // Decimally shift left by `max_decimals - decimals`.
+        let scaleup = max_decimals - decimals.or_else(|| Some(0)).unwrap();
+        for _ in 0..scaleup {
+            value = match INNER_TEN.checked_mul(value) {
+                Some(v) => v,
+                None => return Err(ParseAmountError::TooBig),
+            };
+        }
+
+        if negative {
+            value *= -1;
+        }
+
+        Ok(Amount::from_inner(value))
+    }
+
+    /// Express this Amount as a floating-point value in the given denomination.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn float_denomination(&self, denom: Denomination) -> f64 {
+        let exp = denom.scale() - P::scale();
+        (self.0 as f64) * 10f64.powi(exp)
+    }
+
+    /// Format this Amount in the given denomination.
+    pub fn fmt_denomination(&self, f: &mut fmt::Write, denom: Denomination) -> fmt::Result {
+        let nb_decimals = (P::scale() - denom.scale()) as usize;
+        write!(f, "{:.*} {}", nb_decimals, self.float_denomination(denom), denom)
+    }
+
+    /// Get a formatter string of this Amount in the given denomination,
+    /// followed by the shorthand of the denomination.
+    pub fn to_string_in(&self, denom: Denomination) -> String {
+        let mut buf = String::new();
+        self.fmt_denomination(&mut buf, denom).unwrap();
+        buf
+    }
+
+    /// Get a new Amount by rescaling this one to a new precision.
+    /// Note that some decimals will get lost if the new precision is lower.
+    pub fn as_rescaled<O: Precision>(&self) -> Amount<O> {
```

Comment:
I just stumbled over this PR and I am super excited for `rust-bitcoin` to finally get a proper `Amount` type :)

Regarding this method: As per [Rust naming convention](https://rust-lang-nursery.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv), I think this method should be named `to_rescaled` since it converts from `borrowed` to `owned`.

Reply:
I see your point, but the funcion converts a `borrowed + copy` value to an `owned + copy` value (which isn't covered by the linked table). I guess this means we should change the function signature to `to_rescaled(self) -> Amount`. On the other hand [`std` seems to use `as_` in our scenario](https://doc.rust-lang.org/std/time/struct.Duration.html#method.as_secs) too.

Reply:
Good point!

Maybe the `as/to_*` naming isn't too appropriate here anyway since it is converting to an `Amount` again.

Maybe `as_rescaled` and `into_rescaled` can be replaced with a single `fn rescale_to<O>(self) -> Amount<O>` function? Would read quite nicely IMO:

```rust
let amount = Amount::from_sat(100_000_000);
let milli_satoshi = amount.rescale_to::<MilliSatoshi>();
```

Since `Amount` is `Copy` anyway, the function can just consume `self` :)

Reply:
Hmm, a single method does make sense. Lost track of my inbox last few days. Making this change tomorrow!